libs/vkd3d/command.c

```diff
@@ -2079,6 +2109,202 @@ static void d3d12_command_list_clear_attachment_inline(struct d3d12_command_list
+            return list->dsv_resource_tracking[i].plane_optimal_mask;
+        }
+    }
+
+    vkd3d_array_reserve((void **)&list->dsv_resource_tracking, &list->dsv_resource_tracking_size,
+            list->dsv_resource_tracking_count + 1, sizeof(*list->dsv_resource_tracking));
+    list->dsv_resource_tracking[list->dsv_resource_tracking_count].resource = resource;
+    list->dsv_resource_tracking[list->dsv_resource_tracking_count].plane_optimal_mask = plane_optimal_mask;
+    list->dsv_resource_tracking_count++;
+    return plane_optimal_mask;
+}
+
+static uint32_t d3d12_command_list_notify_dsv_writes(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, const struct vkd3d_view *view, uint32_t plane_write_mask)
+{
+    assert(!(plane_write_mask & ~(VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL)));
+
+    /* If we cover the entire resource, we can promote it to our target layout. */
+    if (view->info.texture.layer_count == resource->desc.DepthOrArraySize &&
+            resource->desc.MipLevels == 1)
+    {
+        return d3d12_command_list_promote_dsv_resource(list, resource, plane_write_mask);
+    }
+    else
+    {
+        d3d12_command_list_get_depth_stencil_resource_layout(list, resource, &plane_write_mask);
+        return plane_write_mask;
+    }
+}
+
+static void d3d12_command_list_notify_dsv_state(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, D3D12_RESOURCE_STATES state, UINT subresource)
+{
+    /* Need to decide if we should promote or decay or promote DSV optimal state.
+     * We can promote if we know for sure that all subresources are optimal.
+     * If we observe any barrier which leaves this state, we must decay. */
+    bool dsv_optimal = state == D3D12_RESOURCE_STATE_DEPTH_READ ||
+            state == D3D12_RESOURCE_STATE_DEPTH_WRITE;
+
+    if (!dsv_optimal)
+    {
+        d3d12_command_list_notify_decay_dsv_resource(list, resource);
+    }
+    else if (subresource == D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES)
+    {
+        d3d12_command_list_promote_dsv_resource(list, resource,
+                VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL);
+    }
+    else if (resource->desc.MipLevels == 1 && resource->desc.DepthOrArraySize == 1)
+    {
+        /* For single mip/layer images (common case for depth-stencil),
+         * a specific subresource can be handled correctly. */
+        if (subresource == 0)
+        {
+            if (resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_DEPTH_BIT)
+                d3d12_command_list_promote_dsv_resource(list, resource, VKD3D_DEPTH_PLANE_OPTIMAL);
+            else if (resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_STENCIL_BIT)
+                d3d12_command_list_promote_dsv_resource(list, resource, VKD3D_STENCIL_PLANE_OPTIMAL);
+        }
+        else if (subresource == 1)
+        {
+            if (resource->format->vk_aspect_mask == (VK_IMAGE_ASPECT_STENCIL_BIT | VK_IMAGE_ASPECT_STENCIL_BIT))
```

Comment:
2x `VK_IMAGE_ASPECT_STENCIL_BIT`, one of these neds to be `VK_IMAGE_ASPECT_DEPTH_BIT`

Reply:
Derp